lib/hanami/cli/commands/app/db/rollback.rb

```
@@ -0,0 +1,175 @@
+                  end
+
+                  next unless dump && !re_running_in_test?
+
+                  run_command(
+                    Structure::Dump,
+                    app: app, slice: slice, gateway: gateway,
+                    command_exit: command_exit
+                  )
+                end
+              end
+            end
+
+            private
+
+            def rollback_across_all_databases(steps:, dump:, gateway:, command_exit:)
+              all_databases = databases(app: false, slice: nil, gateway: gateway)
+
+              # Collect all applied migrations across all databases with their source database
+              all_migrations = []
+              all_databases.each do |database|
+                applied_migrations = database.applied_migrations
+                applied_migrations.each do |migration|
+                  timestamp = Integer(migration.split("_").first)
+                  all_migrations << {
+                    timestamp: timestamp,
+                    name: File.basename(migration, ".*"),
+                    database: database,
+                    migration: migration
+                  }
+                end
+              end
+
+              all_migrations.sort_by! { |m| -m[:timestamp] }
+              migrations_to_rollback = all_migrations.take(steps)
+
+              if migrations_to_rollback.empty?
+                out.puts "==> no migrations to rollback"
+                return
+              end
+
+              migrations_to_rollback.each do |migration_info|
+                database = migration_info[:database]
+                migration_name = migration_info[:name]
+
+                # Find the previous migration in this database
+                all_db_migrations = database.applied_migrations
+                current_index = all_db_migrations.index { |m| m.include?(migration_info[:migration]) }
+
+                next if current_index.nil? # This shouldn't happen, but I am not like 100% sure?
```

i agree i don't think this should happen, _but_ i like keeping here, better safe than sorry! :smile: 

could this even be some sort of error message instead of silently skipping? would that be useful, or exposing internal implementation too much?

---

unsure either, so I left it as `next` but if someone sees a strong incentive to have it as error I will change it :+1: